File: app/uno_gpio_ws/gpio_example.c
#include <stdio.h>
#include <gpiod.h>
#include <unistd.h>

//需安裝 libgpiod-dev 套件 (sudo apt install libgpiod-dev) 和 gpiod 套件 (sudo apt install gpiod)
//編譯指令: gcc -o gpio_example gpio_example.c -lgpiod

#define GPIOCHIP "/dev/gpiochip0"

// DI / DO GPIO 代號
const int DI_PINS[8] = {17, 19, 20, 21, 22, 23, 24, 18}; // DI0 ~ DI7
const int DO_PINS[8] = {25, 26, 27, 28, 29, 30, 31, 32}; // DO0 ~ DO7

// 自訂 GPIO 名稱
const char *custom_gpio_names[8 + 8] = {
    "DI0", "DI1", "DI2", "DI3", "DI4", "DI5", "DI6", "DI7",
    "DO0", "DO1", "DO2", "DO3", "DO4", "DO5", "DO6", "DO7"
};

void getGpioName() {
    struct gpiod_chip *chip;
    struct gpiod_line *line;
    const char *name;

    chip = gpiod_chip_open(GPIOCHIP);
    if (!chip) {
        perror("Failed to open GPIO chip");
        return;
    }

    printf("GPIO Line Information:\n");

    // 查詢 DI / DO 的名稱
    for (int i = 0; i < 8; i++) {
        // **查詢 DI**
        line = gpiod_chip_get_line(chip, DI_PINS[i]);
        name = gpiod_line_name(line);
        printf("GPIO %d: %s (Custom: %s)\n", DI_PINS[i], name ? name : "(Unnamed)", custom_gpio_names[i]);
    }

    for (int i = 0; i < 8; i++) {
        // **查詢 DO**
        line = gpiod_chip_get_line(chip, DO_PINS[i]);
        name = gpiod_line_name(line);
        printf("GPIO %d: %s (Custom: %s)\n", DO_PINS[i], name ? name : "(Unnamed)", custom_gpio_names[i + 8]);
    }

    gpiod_chip_close(chip);
}


int main() {
    struct gpiod_chip *chip;
    struct gpiod_line *di_lines[8], *do_lines[8];
    int values[8];
    
    getGpioName();

    // 開啟 GPIO 晶片
    chip = gpiod_chip_open(GPIOCHIP);
    if (!chip) {
        perror("Failed to open GPIO chip");
        return 1;
    }

    // 設定 DI (輸入)
    for (int i = 0; i < 8; i++) {
        di_lines[i] = gpiod_chip_get_line(chip, DI_PINS[i]);
        if (!di_lines[i] || gpiod_line_request_input(di_lines[i], "gpio_example") < 0) {
            perror("Failed to set DI as input");
            gpiod_chip_close(chip);
            return 1;
        }
    }

    // 設定 DO (輸出)
    for (int i = 0; i < 8; i++) {
        do_lines[i] = gpiod_chip_get_line(chip, DO_PINS[i]);
        if (!do_lines[i] || gpiod_line_request_output(do_lines[i], "gpio_example", 0) < 0) {
            perror("Failed to set DO as output");
            gpiod_chip_close(chip);
            return 1;
        }
    }

    while (1) {
        // 讀取 DI 狀態
        printf("DI: ");
        for (int i = 0; i < 8; i++) {
            values[i] = gpiod_line_get_value(di_lines[i]);
            printf("%d ", values[i]);
        }
        printf("\n");

        // 設定 DO (輸出 DI 的反相值)
        for (int i = 0; i < 8; i++) {
            gpiod_line_set_value(do_lines[i], values[i]); // 反相輸出
        }

        sleep(1); // 每秒更新一次
    }

    // 釋放 GPIO
    for (int i = 0; i < 8; i++) {
        gpiod_line_release(di_lines[i]);
        gpiod_line_release(do_lines[i]);
    }
    gpiod_chip_close(chip);

    return 0;
}

File: app/uno_gpio_ws/gpio_example.py
import gpiod
import time

#需要安裝 python3-libgpiod 和 gpiod 套件
# sudo apt-get install python3-libgpiod
# sudo apt-get install gpiod 

GPIOCHIP = "/dev/gpiochip0"

# DI / DO GPIO 代號
DI_PINS = [17, 19, 20, 21, 22, 23, 24, 18]  # DI0 ~ DI7
DO_PINS = [25, 26, 27, 28, 29, 30, 31, 32]  # DO0 ~ DO7

# 自訂 GPIO 名稱
CUSTOM_GPIO_NAMES = [
    "DI0", "DI1", "DI2", "DI3", "DI4", "DI5", "DI6", "DI7",
    "DO0", "DO1", "DO2", "DO3", "DO4", "DO5", "DO6", "DO7"
]


def get_gpio_names():
    """ 取得 GPIO 名稱 """
    chip = gpiod.Chip(GPIOCHIP)
    
    print("GPIO Line Information:")
    
    # 查詢 DI / DO 的名稱
    for i, pin in enumerate(DI_PINS):
        line = chip.get_line(pin)
        name = line.name() or "(Unnamed)"
        print(f"GPIO {pin}: {name} (Custom: {CUSTOM_GPIO_NAMES[i]})")

    for i, pin in enumerate(DO_PINS):
        line = chip.get_line(pin)
        name = line.name() or "(Unnamed)"
        print(f"GPIO {pin}: {name} (Custom: {CUSTOM_GPIO_NAMES[i + 8]})")


def main():
    """ 主程式 """
    get_gpio_names()

    chip = gpiod.Chip(GPIOCHIP)

    # 設定 DI (輸入)
    di_lines = [chip.get_line(pin) for pin in DI_PINS]
    for line in di_lines:
        line.request(consumer="gpio_example", type=gpiod.LINE_REQ_DIR_IN)

    # 設定 DO (輸出)
    do_lines = [chip.get_line(pin) for pin in DO_PINS]
    for line in do_lines:
        line.request(consumer="gpio_example", type=gpiod.LINE_REQ_DIR_OUT, default_vals=[0])

    while True:
        # 讀取 DI 狀態
        values = [line.get_value() for line in di_lines]
        print("DI:", " ".join(map(str, values)))

        # 設定 DO (輸出 DI 的反相值)
        for line, value in zip(do_lines, values):
            line.set_value(value)  # 直接輸出值

        time.sleep(1)  # 每秒更新一次


if __name__ == "__main__":
    main()
